diff --git a/node_modules/eslint-plugin-lodash-fp/rules/preferred-alias.js b/node_modules/eslint-plugin-lodash-fp/rules/preferred-alias.js
index 24b89f7..b2de3ec 100644
--- a/node_modules/eslint-plugin-lodash-fp/rules/preferred-alias.js
+++ b/node_modules/eslint-plugin-lodash-fp/rules/preferred-alias.js
@@ -32,7 +32,7 @@ const create = function (context) {
   return info.merge({
     CallExpression(node) {
       const method = info.helpers.isMethodCall(node);
-      if (!method) {
+      if (!method || !_.has(data.realToAlias, method.realName)) {
         return;
       }
       const alias = wantedAlias(overrides, method);
